File: backend/app/services/llm/openrouter.py
from __future__ import annotations

import json
from typing import AsyncIterator

import httpx

from app.core.config import settings
from app.services.llm.base import LLMProvider, LLMStreamChunk


class OpenRouterProvider(LLMProvider):
    """Stream completions from OpenRouter's OpenAI-compatible API."""

    name = "openrouter"

    def __init__(self) -> None:
        if not settings.openrouter_api_key:
            raise RuntimeError(
                "OPENROUTER_API_KEY is required when AI_PROVIDER=openrouter"
            )

    async def stream(
        self,
        *,
        system: str | None,
        prompt: str,
        model: str,
    ) -> AsyncIterator[LLMStreamChunk]:
        messages: list[dict] = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        payload = {
            "model": model or settings.ai_model,
            "messages": messages,
            "stream": True,
        }
        headers = {
            "Authorization": f"Bearer {settings.openrouter_api_key}",
            "Content-Type": "application/json",
        }
        url = settings.openrouter_base_url.rstrip("/") + "/chat/completions"
        timeout = httpx.Timeout(60.0, read=300.0)

        async with httpx.AsyncClient(timeout=timeout) as client:
            async with client.stream("POST", url, json=payload, headers=headers) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if not line or not line.startswith("data:"):
                        continue
                    data_str = line[len("data:"):].strip()
                    if data_str == "[DONE]":
                        yield LLMStreamChunk(content="", done=True)
                        return
                    try:
                        data = json.loads(data_str)
                    except json.JSONDecodeError:
                        continue
                    choices = data.get("choices") or []
                    if not choices:
                        continue
                    delta = choices[0].get("delta") or {}
                    content = delta.get("content")
                    if content:
                        yield LLMStreamChunk(content=content)

File: src/lib/ai/openrouter.ts
import "server-only";
import { MODELS } from "./models";
import { recordUsage } from "./usage";

const BASE = "https://openrouter.ai/api/v1";

function apiKey(): string {
  const key = process.env.OPENROUTER_API_KEY;
  if (!key) throw Object.assign(new Error("AI is not configured"), { status: 503 });
  return key;
}

const HEADERS = () => ({
  Authorization: `Bearer ${apiKey()}`,
  "Content-Type": "application/json",
  "HTTP-Referer": "https://veridoc.app",
  "X-Title": "Veridoc",
});

export type ChatMessage = { role: "system" | "user" | "assistant"; content: string };

export type ChatOptions = {
  model: string;
  messages: ChatMessage[];
  temperature?: number;
  maxTokens?: number;
  workspaceId?: string;
  purpose?: string;
};

/** Non-streaming chat completion. Records token usage when a workspace is given. */
export async function chat(opts: ChatOptions): Promise<string> {
  const res = await fetch(`${BASE}/chat/completions`, {
    method: "POST",
    headers: HEADERS(),
    body: JSON.stringify({
      model: opts.model,
      messages: opts.messages,
      temperature: opts.temperature ?? 0.2,
      max_tokens: opts.maxTokens ?? 1024,
    }),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw Object.assign(new Error(`AI request failed: ${res.status} ${text.slice(0, 200)}`), {
      status: 502,
    });
  }
  const data = await res.json();
  const usage = data.usage ?? {};
  if (opts.workspaceId) {
    await recordUsage({
      workspaceId: opts.workspaceId,
      model: opts.model,
      purpose: opts.purpose ?? "chat",
      tokensIn: usage.prompt_tokens ?? 0,
      tokensOut: usage.completion_tokens ?? 0,
    }).catch(() => {});
  }
  return data.choices?.[0]?.message?.content ?? "";
}

/** Streaming chat completion as an async iterator of text deltas. */
export async function* chatStream(opts: ChatOptions): AsyncGenerator<string> {
  const res = await fetch(`${BASE}/chat/completions`, {
    method: "POST",
    headers: HEADERS(),
    body: JSON.stringify({
      model: opts.model,
      messages: opts.messages,
      temperature: opts.temperature ?? 0.2,
      max_tokens: opts.maxTokens ?? 1024,
      stream: true,
    }),
  });
  if (!res.ok || !res.body) {
    const text = await res.text().catch(() => "");
    throw Object.assign(new Error(`AI stream failed: ${res.status} ${text.slice(0, 200)}`), {
      status: 502,
    });
  }

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let tokensOut = 0;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith("data:")) continue;
      const payload = trimmed.slice(5).trim();
      if (payload === "[DONE]") continue;
      try {
        const json = JSON.parse(payload);
        const delta = json.choices?.[0]?.delta?.content;
        if (delta) {
          tokensOut += 1;
          yield delta as string;
        }
      } catch {
        // ignore keep-alive / partial frames
      }
    }
  }

  if (opts.workspaceId) {
    await recordUsage({
      workspaceId: opts.workspaceId,
      model: opts.model,
      purpose: opts.purpose ?? "copilot",
      tokensIn: 0,
      tokensOut,
    }).catch(() => {});
  }
}

/** Embeds an array of texts. Returns one vector per input. */
export async function embed(
  texts: string[],
  workspaceId?: string
): Promise<number[][]> {
  if (texts.length === 0) return [];
  const res = await fetch(`${BASE}/embeddings`, {
    method: "POST",
    headers: HEADERS(),
    body: JSON.stringify({ model: MODELS.embed, input: texts }),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw Object.assign(new Error(`Embedding failed: ${res.status} ${text.slice(0, 200)}`), {
      status: 502,
    });
  }
  const data = await res.json();
  if (workspaceId) {
    await recordUsage({
      workspaceId,
      model: MODELS.embed,
      purpose: "embed",
      tokensIn: data.usage?.prompt_tokens ?? 0,
      tokensOut: 0,
    }).catch(() => {});
  }
  return (data.data ?? []).map((d: { embedding: number[] }) => d.embedding);
}
